File: 17.Backend Security - Everything You Need to Know/code/typescript/slow_hashing_defeating_brute_force_3.ts
import * as crypto from "crypto";
import { promisify } from "util";

const randomBytes = promisify(crypto.randomBytes);

export async function hashPassword(password: string): Promise<string> {
  // Generate cryptographically random 16-byte salt
  const salt = await randomBytes(16);

  // Argon2id params equivalent for Node's scrypt or using a package.
  // We'll use scrypt for standard node built-ins, or we can use argon2 package.
  return new Promise((resolve, reject) => {
    crypto.scrypt(password, salt, 64, { N: 16384, r: 8, p: 1 }, (err, derivedKey) => {
      if (err) return reject(err);
      
      // Store: $salt$hash (both needed to verify)
      const encoded = `${salt.toString("base64")}$${derivedKey.toString("base64")}`;
      resolve(encoded);
    });
  });
}

export async function verifyPassword(password: string, stored: string): Promise<boolean> {
  // Re-hash with stored salt, compare, never compare raw hashes with ==
  const [saltB64, hashB64] = stored.split("$");
  const salt = Buffer.from(saltB64, "base64");
  const expectedHash = Buffer.from(hashB64, "base64");

  return new Promise((resolve, reject) => {
    crypto.scrypt(password, salt, 64, { N: 16384, r: 8, p: 1 }, (err, derivedKey) => {
      if (err) return reject(err);
      // Use timingSafeEqual to prevent timing attacks
      resolve(crypto.timingSafeEqual(expectedHash, derivedKey));
    });
  });
}

File: 17.Backend Security - Everything You Need to Know/code/rust/slow_hashing_defeating_brute_force_3.rs
use argon2::{
    password_hash::{rand_core::OsRng, PasswordHash, PasswordHasher, PasswordVerifier, SaltString},
    Argon2,
};

pub fn hash_password(password: &str) -> Result<String, argon2::password_hash::Error> {
    // Generate cryptographically random 16-byte salt
    let salt = SaltString::generate(&mut OsRng);

    // Argon2id params: default handles time, memory, threads
    let argon2 = Argon2::default();

    // Store: $argon2id$v=19$m=4096,t=3,p=1$salt$hash (both needed to verify)
    let password_hash = argon2.hash_password(password.as_bytes(), &salt)?.to_string();
    Ok(password_hash)
}

pub fn verify_password(password: &str, stored: &str) -> bool {
    // Re-hash with stored salt, compare, never compare raw hashes with ==
    let parsed_hash = match PasswordHash::new(stored) {
        Ok(h) => h,
        Err(_) => return false,
    };
    
    Argon2::default()
        .verify_password(password.as_bytes(), &parsed_hash)
        .is_ok()
}
